driver-core/src/main/com/mongodb/internal/connection/IdHoldingBsonWriter.java

```diff
@@ -127,15 +127,14 @@ public void writeStartArray(final String name) {
 
     @Override
     public void writeEndArray() {
-        super.writeEndArray();
-
         if (isWritingId()) {
             getIdBsonWriter().writeEndArray();
             if (getIdBsonWriter().getCurrentLevel() == 0) {
                 getIdBsonWriter().writeEndDocument();
                 id = new RawBsonDocument(getBytes()).get(ID_FIELD_NAME);
             }
         }
+        super.writeEndArray();
```

Comment:
Intuitively it makes sense that if `writeStartArray` calls `super.writeStartArray` at the end, that `writeEndArray` would call it at the beginning.  Can you explain why that intuition is incorrect?

Reply:
This should be similar to the `writeEndDocument` method starting on line 82.   The order of operations in `writeStartArray` should be followed. First, check if the last field written is `_id` and decrement the level in the IdHoldingBsonWriter. Then call `super.writeEndArray` to decrement the level in the LevelCountingBsonWriter.